File: projects/07-Treaps/implementation/treaps.java
import java.util.Random;

public class treaps {
    /*
    a treap is a binary search tree where each node also has a random priority.
    the tree is heap ordered by priority, while also maintaining the binary search tree property by key.
    insertions and deletions are done by splitting and merging the treap based on the priorities of the nodes.
    this allows for efficient insertions, deletions, and searches while keeping the tree balanced on average.
    */

    // we need a random number generator to assign priorities to the nodes
    private static final Random rand = new Random();

    // we simply define a node to represent the nodes in the treap
    static class Node {
        int key;
        int priority;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.priority = rand.nextInt();
        }
    }

    private Node root;

    /*
    the split function takes a node and a key and will split the treap
    into two: one where all the elements are greater than the key we are at and one where
    the elements are less than or equal to the key we are at.
    the way it works is when we are at a node, if the key is less than or equal to the current nodes value,
    we split the nodes right subtree because there may be larger values that we need to split,
    if the key is greater than the current nodes value, we split the nodes left subtree because there may be smaller values that we need to split.
    so we will create two treaps both priority ordered, split by key, and we will return them as an array of nodes.
    */
    private Node[] split(Node root, int key) {
        if (root == null) {
            return new Node[]{null, null};
        }
        // less than or equal to the key, we split the right subtree
        if (root.key <= key) {
            Node[] res = split(root.right, key);
            // we want to set the root.right to be all the values in the right subtree that are less than or equal to the key we are at
            // this allows us to make sure none of the wrong values are in the wrong treap
            root.right = res[0];
            // we return the root as the left treap and res[1] as the right treap
            // so when we recurse back up, we will have the correct treaps with the correct priorities and keys
            return new Node[]{root, res[1]};
        } else {
            Node[] res = split(root.left, key);
            // similar logic to the above case, just simply reversed because we are splitting the left subtree instead of the right subtree

            root.left = res[1];

            return new Node[]{res[0], root};
        }
    }

    /*
    the point of the merge function is to take two treaps and merge them together
    in a manner that maintains the properties of the treap,
    so we will compare the priorities of the roots of the two treaps and merge the one with the higher priority as the new root,
    and then we will merge the other treap as a child of the new root,
    we will do this recursively until we have merged the two treaps together
    */
    private Node merge(Node left, Node right) {
        if (left == null) return right;
        if (right == null) return left;

        // the left treap has a higher priority than the right so we will call a merge on the left treaps right child
        // since that may have a higher priority than the right treap and we will return the left treap as the new root
        if (left.priority > right.priority) {
            left.right = merge(left.right, right);
            return left;
        // if the right priority is higher, choose the right root
        } else {
            right.left = merge(left, right.left);
            return right;
        }
    }

    /*
    this insert is straightforward, so if we hit the case in our recursion that the node were on's
    priority is less than the inserted node, we will split the current node by the inserted node's key and make the inserted node the new root of the subtree,
    and then we will set the left and right children of the new root to be the two treaps that we got from the split function,
    this allows us to insert our node while maintaining treap property
    */
    // we call insert separately because we need to create a new node

    public void insert(int key) {
        root = insert(root, new Node(key));
    }

    private Node insert(Node root, Node node) {
        if (root == null) return node;
        // check if node priority is > root priority,
        // if it is we call our split function and place node at current root
        if (node.priority > root.priority) {
            Node[] halves = split(root, node.key);

            node.left = halves[0];
            node.right = halves[1];

            return node;
        }
        // traverse the tree to maintain binary search tree order
        if (node.key < root.key) {
            root.left = insert(root.left, node);
        } else if (node.key > root.key) {
            root.right = insert(root.right, node);
        }
        return root;
    }

    /*
    the delete function is easy because all we do is merge
    the left and right subtrees of the node we want to delete,
    this merge will maintain the treap properties and we will return the new root of the merged treap,
    skipping over the node we want to delete, and then return all the way back up with the new treap
    */

    // again call it separately, same reason as insert
    public void delete(int key) {
        root = delete(root, key);
    }

    private Node delete(Node root, int key) {
        if (root == null) return null;

        // if found, remove node by merging its children
        if (root.key == key) {
            return merge(root.left, root.right);
        }

        // traverse to locate node, then refresh pointer
        if (key < root.key) {
            root.left = delete(root.left, key);
        } else {
            root.right = delete(root.right, key);
        }

        return root;
    }

    // search for key in treap (BST traversal)
    public boolean search(int key) {
        return search(root, key);
    }

    private boolean search(Node root, int key) {
        // normal search
        if (root == null) return false;
        if (root.key == key) return true;
        if (key < root.key) {
            return search(root.left, key);
        } else {
            return search(root.right, key);
        }
    }

    // inorder traversal for sorted output (by key)
    public void inorder() {
        inorder(root);
        System.out.println();
    }

    private void inorder(Node root) {
        if (root == null) return;
        inorder(root.left);
        System.out.print(root.key + " ");
        inorder(root.right);
    }

    // this function is just to help us test because 
    // our current inorder function only prints the values but we need a comparison
    private static String inorderString(treaps t) {
        return inorderString(t.root).trim();
    }

    private static String inorderString(Node root) {
        if (root == null) return "";
        String left = inorderString(root.left);
        String node = Integer.toString(root.key);
        String right = inorderString(root.right);

        String result = "";
        if (!left.isEmpty()) {
            result += left + " ";
        }; 
        result += node;
        if (!right.isEmpty()) result += " " + right;
        return result;
    }

    /*
    
    */
    public static void main(String[] args) {
        int passed = 0;
        int total = 4;

        // basic insert and search
        treaps treaps = new treaps();
        treaps.insert(5);
        treaps.insert(3);
        treaps.insert(7);
        treaps.insert(2);
        treaps.insert(4);
        boolean testinsert = treaps.search(5) && treaps.search(2) && !treaps.search(6);
        System.out.println("Test 1: insert/search is a" + (testinsert ? "PASS" : "FAIL"));
        if (testinsert) {
            passed++;
        }

        // make sure inorder is right
        // test inorder should be 2 3 4 5 7 because we inserted those values 
        String expected = "2 3 4 5 7";
        String actual = inorderString(treaps);
        boolean testinorder = expected.equals(actual);
        System.out.println("Test 2: inorder is a " + (testinorder ? "PASS" : "FAIL"));
        if (testinorder) {
            passed++;
        }


        // delete
        // test that we went from 2 3 4 5 7 to 2 4 5 7 and that search for the deleted node returns false
        treaps.delete(3);
        String expecteddel = "2 4 5 7";
        String actualdel = inorderString(treaps);
        boolean testdel = expecteddel.equals(actualdel) && !treaps.search(3);
        System.out.println("Test 3: delete is a " + (testdel ? "PASS" : "FAIL"));
        if (testdel) {
            passed++; 
        } 

        // Test Case 4: if we try to delete something that doesnt exist it wont change 
        treaps.delete(42);
        String expected4 = "2 4 5 7";
        String actual4 = inorderString(treaps);
        boolean test4 = expected4.equals(actual4);
        System.out.println("Test 4: delete something that doesnt exist  " + (test4 ? "PASS" : "FAIL") + " (got: " + actual4 + ")");
        if (test4) {
            passed++;
        }
        // test that all of our cases passed and return to user!
        System.out.println("\nwhen running my treaps you got: " + passed + " / " + total + " tests passed.");
        if (passed != total) {
            System.exit(1);
        }
    }
}

File: projects/07-Treaps/implementation/treaps.py
import random

"""
Treap (tree + heap) - a randomized binary search tree with heap properties
a treap combines BST ordering with random priorities to maintain balance
"""

# a simple node in a treap structure
class TreapNode:
    def __init__(self, key):
        self.key = key

        # random priority for every node in order to maintain the heap property
        self.priority = random.random()  
        self.left = None
        self.right = None

# the actual treap data structure implementation
class Treap:
    
    def __init__(self):
        self.root = None

    """
    splitting the treap into two parts based on the key using simple recursion
    the left treap will contain all keys <= key and the right treap will contain all keys > key
    this is a helper function for insertion and deletion to maintain the treap properties
    """
    def split(self, root, key):
        if root is None:
            return (None, None)
        """
        if the root's key is less than or equal to the key, then the split must be in the right subtree and the resulting left treap 
        will include the original left subtree as well as a portion of the original right subtree.

        if the root's key is greater than the key, then the split must be in the left subtree and the resulting right treap will include
        the original right subtree as well as a portion of the original left subtree.
        """
        if root.key <= key:
            root.right, right = self.split(root.right, key)
            return (root, right)
        else:
            left, root.left = self.split(root.left, key)
            return (left, root)
    
    """
    merge two treaps and return the resulting treap. 
    merge operation is based on the priorities of the root nodes of the two treaps since
    the treap with the higher priority root becomes the new root of the merged treap, and the other treap is merged as a subtree of the new root. 
    """
    def merge(self, left, right):
        if left is None:
            return right
        if right is None:
            return left
        
        # here we recursively merge the treaps based on their priorities
        if left.priority > right.priority:
            left.right = self.merge(left.right, right)
            return left
        else:
            right.left = self.merge(left, right.left)
            return right

    """
    inserting a key into the treap is done by first splitting the treap into two parts based on the key, then creating a new node with the key and merging it with the two parts.
    this ensures that the new node is placed in the correct position according to the BST property, and the merge operation maintains the heap property of the treap.
    this is indeed a different approach and implementation of insert, yet it achieves the insertion in a less BST-like manner
    """
    def insert(self, key):
        if self.search(key):
            return

        new_node = TreapNode(key)
        left, right = self.split(self.root, key)
        temp = self.merge(left, new_node)
        self.root = self.merge(temp, right)

    """
    removes a key by finding the node and replacing it with a merge of its children.
    instead of manually restructuring, we let merge decide the new root based on priorities.
    """
    def erase(self, key):
        self.root = self._erase_helper(self.root, key)
    
    """
    walks down the tree like a normal BST search until the target node is found.
    once found, the node is removed by merging its left and right subtrees together,
    effectively "skipping over" the node while keeping the tree connected.
    """
    def _erase_helper(self, node, key):
        if node is None:
            return None
        
        # move left or right depending on where the key should be
        if key < node.key:
            node.left = self._erase_helper(node.left, key)
        elif key > node.key:
            node.right = self._erase_helper(node.right, key)
        else:
            # found the node: replace it by merging its two children
            return self.merge(node.left, node.right)
        
        return node 

    """
    combines two treaps into one by recursively choosing which root should come first.
    the structure is rebuilt by always keeping the higher-priority node above.
    """
    def unite(self, other):
        self.root = self._unite_helper(self.root, other.root)

    """
    compares the roots of both trees and picks the one that should be higher.
    then recursively attaches the remaining part of the other tree into the correct subtree.
    """
    def _unite_helper(self, node1, node2):
        if node1 is None:
            return node2
        if node2 is None:
            return node1

        if node1.priority < node2.priority:
            node1, node2 = node2, node1

        left2, right2 = self.split(node2, node1.key)
        node1.left = self._unite_helper(node1.left, left2)
        node1.right = self._unite_helper(node1.right, right2)
        return node1
    
    """
    checks whether a key exists by following the same path a BST would take.
    priorities are irrelevant here since they only affect structure, not lookup direction.
    """
    def search(self, key):
        return self._search_helper(self.root, key)
    
    """
    at each step, decide whether to go left or right based on the key.
    stop early if the value is found, or return false if a leaf is reached.
    """
    def _search_helper(self, node, key):
        if node is None:
            return False
        
        if key == node.key:
            return True
        elif key < node.key:
            return self._search_helper(node.left, key)
        else:
            return self._search_helper(node.right, key)

    """
    returns all keys in sorted order by performing an inorder traversal.
    this visits nodes in increasing order of their keys, and this is useful for debugging
    """
    def inorder(self):
        result = []
        self._inorder_helper(self.root, result)
        return result
    
    """
    standard inorder traversal: left subtree, current node, right subtree.
    values are appended as we visit each node.
    """
    def _inorder_helper(self, node, result):
        if node is not None:
            self._inorder_helper(node.left, result)
            result.append(node.key)
            self._inorder_helper(node.right, result)


def inorder_string(treap):
        return " ".join(str(x) for x in treap.inorder())


if __name__ == "__main__":
    passed = 0
    total = 4

    # basic insert and search
    treaps = Treap()
    treaps.insert(5)
    treaps.insert(3)
    treaps.insert(7)
    treaps.insert(2)
    treaps.insert(4)

    test_insert = treaps.search(5) and treaps.search(2) and not treaps.search(6)
    print("Test 1: insert/search is a", "PASS" if test_insert else "FAIL")
    if test_insert:
        passed += 1

    # test inorder
    expected = "2 3 4 5 7"
    actual = inorder_string(treaps)
    test_inorder = expected == actual
    print("Test 2: inorder is a", "PASS" if test_inorder else "FAIL")
    if test_inorder:
        passed += 1

    # delete test (your method is called erase, not delete)
    treaps.erase(3)
    expected_del = "2 4 5 7"
    actual_del = inorder_string(treaps)
    test_del = expected_del == actual_del and not treaps.search(3)
    print("Test 3: delete is a", "PASS" if test_del else "FAIL")
    if test_del:
        passed += 1

    # delete non-existent
    treaps.erase(42)
    expected4 = "2 4 5 7"
    actual4 = inorder_string(treaps)
    test4 = expected4 == actual4
    print("Test 4: delete something that doesnt exist", "PASS" if test4 else "FAIL", f"(got: {actual4})")
    if test4:
        passed += 1

    print(f"\nwhen running my treaps you got: {passed} / {total} tests passed.")
    if passed != total:
        exit(1)